ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto

```diff
@@ -128,6 +135,74 @@ message GetFlatTransactionResponse {
   Transaction transaction = 1;
 }
 
+message GetEventsByContractIdRequest {
+  // The contract id being queried.
+  // Required
+  string contract_id = 1;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract id will only be returned if the requesting parties includes
+  // at least one party that is a witness of the action that created the event. For a full description
+  // of witness see https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#projections.
+  // Required
+  repeated string requesting_parties = 2;
+
+}
+message GetEventsByContractIdResponse {
+  // The collection of events, will only contain CreateEvent and consuming ExerciseEvents. If the
+  // requested `contract_id` has been archived or never existed this collection will be empty.
+  repeated TreeEvent events = 1;
+}
+
+message GetEventsByContractKeyRequest {
+
+  // The contract key to search for.
+  // Required
+  Value contract_key = 1;
+
+  // The template id associated with the contract key
+  // Required
+  Identifier template_id = 2;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract key will only be returned if the requesting parties includes
+  // at least one party that is a witness of the action that created the event. For a full description
+  // of witness see https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#projections.
+  // Required
+  repeated string requesting_parties = 3;
+
+  // Optional - will default to 1000.
+  uint32 max_events = 4;
```

Comment:
The reason we introduced pagination and did not use a stream  is:
- The expected use case will only return a small number of rows
- As this is used for investigation easier to work with via grpcurl or canton console.
- However one can see there may be a use case where 000s of contracts are returned which is where the pagination covers us.
- This was also done because it is quite possible that we will want to include non-consuming events in the future.